pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,296 @@
+		}
+
+		listAction := action.NewList(actionConfig)
+		releases, err := listAction.Run()
+		if err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to list Helm releases for namespace %v", namespace)
+		}
+
+		for _, helmRelease := range releases {
+			lastRelease, err := actionConfig.Releases.Last(helmRelease.Name)
+			if err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to find latest release version for release %v", helmRelease.Name)
+				// If this fails, something went wrong. Skip to the next release.
+				continue
+			}
+
+			// TODO consume the function from helm-mapkubeapis once that is merged in
+			replaced, modifiedManifest, err := ReplaceManifestData(apiMappings, lastRelease.Manifest, cluster.Status.Version.GitVersion)
+			if err != nil {
+				// If this fails, it probably means we don't have adequate write permissions
+				return errors.Wrapf(err, "[cleanupHelmReleases] failed to replace deprecated/removed APIs on cluster %v", cluster.Name)
+			}
+
+			if !replaced {
+				logrus.Infof("[cleanupHelmReleases] release %v in namespace %v has no deprecated or removed APIs", lastRelease.Name, namespace)
+				continue
+			}
+
+			if err := updateRelease(lastRelease, modifiedManifest, actionConfig); err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to update release %v in namespace %v, skipping...", lastRelease.Name, lastRelease.Namespace)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ReplaceManifestData replaces the out-of-date APIs with their respective valid successors, or removes an API that
+// does not have a successor.
+// Logic extracted from https://github.com/stormqueen1990/helm-mapkubeapis/blob/0245b7a7837a36fd164d83e496c453811d62c083/pkg/common/common.go#L81-L142
+func ReplaceManifestData(mapMetadata *mapping.Metadata, manifest string, kubeVersion string) (bool, string, error) {
+	var replaced = false
+	for _, mappingData := range mapMetadata.Mappings {
+		deprecatedAPI := mappingData.DeprecatedAPI
+		supportedAPI := mappingData.NewAPI
+		var apiVersion string
+
+		if mappingData.DeprecatedInVersion != "" {
+			apiVersion = mappingData.DeprecatedInVersion
+		} else {
+			apiVersion = mappingData.RemovedInVersion
+		}
+
+		if !semver.IsValid(kubeVersion) {
+			return replaced, "", errors.Errorf("Invalid format for Kubernetes semantic version: %v", kubeVersion)
+		}
+
+		if !semver.IsValid(apiVersion) {
+			return replaced, "", errors.Errorf("Failed to get the deprecated or removed Kubernetes version for API: %s", strings.ReplaceAll(deprecatedAPI, "\n", " "))
+		}
+
+		if count := strings.Count(manifest, deprecatedAPI); count > 0 {
+			if semver.Compare(apiVersion, kubeVersion) > 0 {
+				logrus.Debugf("The following API does not require mapping as the "+
+					"API is not deprecated or removed in Kubernetes '%s':\n\"%s\"\n", apiVersion,
+					deprecatedAPI)
+			} else {
+				if supportedAPI == "" {
```

Comment:
Looks like in this case "continue" can be used and else removed below.
```suggestion
				logrus.Debugf("The following API does not require mapping as the "+
					"API is not deprecated or removed in Kubernetes '%s':\n\"%s\"\n", apiVersion,
					deprecatedAPI)
				continue
			}
			if supportedAPI == "" {
```

Reply:
Updated!